general/mem_manager.hpp

```diff
@@ -316,6 +318,11 @@ class Memory
        i.e. it will, generally, not be Empty() after this call. */
    inline void Delete();
 
+   /** @brief Delete the owned device pointer. */
+   inline void DeleteDevice();
```

Comment:
I think the description (and the behavior) should be something like this:
"Delete the device pointer, if owned. If the data is valid only on device, move it to host before deleting and invalidating the device memory."

Reply:
Or it maybe more flexible to have an optional boolean parameter `copy_to_host` which can be `true` by default.